Nightly: Added TCP Keepalive test over tunnel

eloycoto opened the pull request:
- Added TCP keepalive over Vxlan without policy and with L7 policy.
- Added ExpectContains helper in resCmd

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Nightly: Added TCP Keepalive test over tunnel

- Added TCP keepalive over Vxlan without policy and with L7 policy.
- Added ExpectContains helper in resCmd

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+			uid = helpers.MakeUID()
+			kubectl.ExecPodCmd(helpers.DefaultNamespace, client,
+				fmt.Sprintf(`echo -e "%s" >> /tmp/nc_pipe.txt`, HTTPRequest(uid)))
+			helpers.Sleep(5) // Give time to fill the buffer in context.
```
How do we know that 5 seconds is enough time?

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+			defer cancel()
+
+			serverctx := kubectl.ExecPodCmdContext(ctx, helpers.DefaultNamespace, server, ncServer)
+			kubectl.ExecPodCmdContext(ctx, helpers.DefaultNamespace, client, ncClient)
```
nit - can you add` _ =` before this so we explicitly know that the return value is being ignored?

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+		}
+
+		It("Test TCP Keepalive with L7 Policy", func() {
+			manifest := kubectl.ManifestGet("netcat_ds.yaml")
```
Make `netcat_ds.yaml` a constant since it's used throughout this file.

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+			By("Sleeping for six  minutes to check tcp-keepalive")
+			helpers.Sleep(360)
+			uid = helpers.MakeUID()
+			kubectl.ExecPodCmd(helpers.DefaultNamespace, client,
```
nit - can you add `_ =` before this so we explicitly know that the return value is being ignored?

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+
+			By("Sleeping for a minute to check tcp-keepalive")
+			helpers.Sleep(60)
+			uid = helpers.MakeUID()
```
This code repeats itself on lines 291-295, and 275-279. Make it into a function.

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+			Expect(err).To(BeNil(), "Cannot get netcat ips")
+
+			ncServer := getServer("8888")
+			ncClient := getClient(ips[server], "8888", "/tmp/nc_pipe.txt")
```
Make the file path a constant since it's reused multiple times throughout the test.

joestringer commented on test/k8sT/Nightly.go:
```diff
@@ -226,6 +227,90 @@ var _ = Describe("NightlyEpsMeasurement", func() {
+
+		getClient := func(ip, port, filePipe string) string {
+			return fmt.Sprintf(
+				"rm %[1]s; touch %[1]s; tail -f %[1]s 2>&1 | nc -v %[2]s %[3]s",
```
FYI there are multiple implementations of `nc` that are used in different distros and they may/may not have the same behaviour with different arguments. In OVS we ended up parameterizing the arguments to pass into `nc` to ensure that it would exit with a five second timeout, and the arguments had to be defined differently on different distros. I'd suggest testing to making sure that if this fails for some reason, that it eventually exits.

https://github.com/openvswitch/ovs/blob/b1db9dd6747cf8962ebcca4fc15dd084728b0f79/tests/atlocal.in#L159-L166

eloycoto committed:
Nightly: Fix some typos

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo: Fixed WaitCleanAllTerminatingEndpoints

Due the test-flake I discovered that the termination helper didn't work
as expected, and the status.phase is not represent at all
(https://github.com/kubernetes/kubernetes/issues/49387)

Issue:

```
vagrant@k8s1:~$ kubectl delete pod testds-w7prl
pod "testds-w7prl" deleted
vagrant@k8s1:~$ kubectl get pods
NAME               READY     STATUS        RESTARTS   AGE
netcatds-bhxv4     1/1       Running       0          5m
netcatds-zpzzl     1/1       Running       0          5m
testclient-8qx59   1/1       Running       0          1m
testclient-r9xmm   1/1       Running       0          1m
testds-fwss5       1/1       Running       0          32s
testds-w7prl       0/1       Terminating   0          1m
vagrant@k8s1:~$ kubectl get pods -o "jsonpath='{.items[*].status.phase}'"
'Running Running Running Running^C
vagrant@k8s1:~$ kubectl get pods
NAME               READY     STATUS        RESTARTS   AGE
netcatds-bhxv4     1/1       Running       0          5m
netcatds-zpzzl     1/1       Running       0          5m
testclient-8qx59   1/1       Running       0          1m
testclient-r9xmm   1/1       Running       0          1m
testds-fwss5       1/1       Running       0          40s
testds-w7prl       0/1       Terminating   0          1m
vagrant@k8s1:~$
```

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>
